python/dgl/heterograph.py

```diff
@@ -0,0 +1,2092 @@
+    def out_edges(self, v, form='uv'):
+        """Return the outbound edges of the node(s).
+
+        Only works if the graph has one edge type.  For multiple types,
+        query with
+
+        .. code::
+
+           g['srctype', 'dsttype', 'edgetype'].edge_ids(u, v)
+
+        Parameters
+        ----------
+        v : int, list, tensor
+            The node(s).
+        form : str, optional
+            The return form. Currently support:
+
+            - 'all' : a tuple (u, v, eid)
+            - 'uv'  : a pair (u, v), default
+            - 'eid' : one eid tensor
+
+        Returns
+        -------
+        A tuple of Tensors ``(eu, ev, eid)`` if ``form == 'all'``.
+            ``eid[i]`` is the ID of an outbound edge from ``eu[i]`` to ``ev[i]``.
+            All outbound edges from ``v`` are returned.
+        A pair of Tensors (eu, ev) if form == 'uv'
+            ``ev[i]`` is the destination node of an outbound edge from ``eu[i]``.
+            All outbound edges from ``v`` are returned.
+        One Tensor if form == 'eid'
+            ``eid[i]`` is ID of an outbound edge from any of the nodes in ``v``.
+
+        Examples
+        --------
+        The following example uses PyTorch backend.
+
+        Find the gameplay IDs of user #0 (Alice)
+        >>> g['user', 'game', 'plays'].out_edges(0, 'eid')
+        tensor([0])
+        """
+        pass
+
+    def all_edges(self, form='uv', order=None):
+        """Return all the edges.
+
+        Only works if the graph has one edge type.  For multiple types,
+        query with
+
+        .. code::
+
+           g['srctype', 'dsttype', 'edgetype'].edge_ids(u, v)
+
+        Parameters
+        ----------
+        form : str, optional
+            The return form. Currently support:
+
+            - 'all' : a tuple (u, v, eid)
+            - 'uv'  : a pair (u, v), default
+            - 'eid' : one eid tensor
+        order : string
+            The order of the returned edges. Currently support:
+
+            - 'srcdst' : sorted by their src and dst ids.
+            - 'eid'    : sorted by edge Ids.
+            - None     : the arbitrary order.
+
+        Returns
+        -------
+        A tuple of Tensors (u, v, eid) if form == 'all'
+            ``eid[i]`` is the ID of an edge between ``u[i]`` and ``v[i]``.
+            All edges are returned.
+        A pair of Tensors (u, v) if form == 'uv'
+            An edge exists between ``u[i]`` and ``v[i]``.
+            If ``n`` edges exist between ``u`` and ``v``, then ``u`` and ``v`` as a pair
+            will appear ``n`` times.
+        One Tensor if form == 'eid'
+            ``eid[i]`` is the ID of an edge in the graph.
+
+        Examples
+        --------
+        The following example uses PyTorch backend.
+
+        Find the user-game pairs for all gameplays:
+        >>> g['user', 'game', 'plays'].all_edges('uv')
+        (tensor([0, 1, 1, 2]), tensor([0, 0, 1, 1]))
+        """
+        pass
+
+    def in_degree(self, v):
```

Comment:
this also require node type.

Reply:
how to get degree in the original heterogeneous graph.

Reply:
> how to get degree in the original heterogeneous graph.

If you want to you can loop over the in-edges of the metagraph, get all the edge types connecting to the node type you're considering, and call this method for each edge type.

And I don't know if the degree of all types is meaningful enough; usually different entities like "User" and "Developer" should not be counted together anyway.

Reply:
I don't know what people want to do. Since this is query API, I feel it should be reasonably flexible. If a hetero graph has 10 node types in `ntypes`. Your solution will be something like this:
```python
in_deg = zeros(...)
for ntype in ntypes:
    in_deg = in_deg + g[ntype].in_degree(v)
for ntype1 in ntypes:
    for ntype2 in ntypes:
        for etype in etypes:
            in_deg = in_deg + g[ntype1, ntype2, etype].in_degree(v)
```
something like this. I think my code is probably incorrect. My point is that users' code can becomes unnecessarily complex and also very slow.

For a bipartite graph, what are `v`? Do they have to be left nodes? right nodes? It has to be one type of nodes, right? It's pretty confusing which side it should be.

Reply:
> I don't know what people want to do. Since this is query API, I feel it should be reasonably flexible. If a hetero graph has 10 node types in `ntypes`. Your solution will be something like this:
> 
> ```python
> in_deg = zeros(...)
> for ntype in ntypes:
>     in_deg = in_deg + g[ntype].in_degree(v)
> for ntype1 in ntypes:
>     for ntype2 in ntypes:
>         for etype in etypes:
>             in_deg = in_deg + g[ntype1, ntype2, etype].in_degree(v)
> ```
> 
> something like this. I think my code is probably incorrect. My point is that users' code can becomes unnecessarily complex and also very slow.

Let's keep the interface simple first, as the current query interface is consistent with `DGLBaseGraph`.  I doubt if people would only care about the total number of predecessors without caring about the types.  If users do want more flexible interfaces we can later add new ones such as `typed_successors`.

> For a bipartite graph, what are `v`? Do they have to be left nodes? right nodes? It has to be one type of nodes, right? It's pretty confusing which side it should be.

They will always be "destination" nodes, since the bipartite graph would be *directed*.  A undirected bipartite graph would be "directionized" and the result would have a relationship (edge type) pointing left to right, and another pointing right to left.  So one can still do
```
g['left', 'right', 'ltor'].in_degrees(...)  # in degrees of right nodes
g['right', 'left', 'rtol'].in_degrees(...)  # in degrees of left nodes
```

Reply:
i see. i feel this is a little counter-intuitive, but it works. we could have users to give us feedback later.